### PR TITLE
[JUJU-820] Improve error message for unspec'd channel

### PR DESCRIPTION
**Pull request by barrettj12**
When deploying a Charmhub charm with specified revision but unspecified channel, Juju was giving a cryptic error message:
```
$ juju deploy ch:hello-juju --revision=3
ERROR invalid channel for "ch:hello-juju": channel cannot be empty
```
I made this message more clear, so the user can tell what to do:
```
$ juju deploy ch:hello-juju --revision=3
ERROR If a revision is specified, a channel must also be specified for Charmhub
charms and bundles. Do this using the --channel option, e.g.
    juju deploy <charm> --revision=5 --channel=stable
```

## Checklist

 - [ ] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [ ] Comments answer the question of why design decisions were made

## QA steps

```sh
make go-install
juju destroy-controller overlord --destroy-all-models
juju bootstrap localhost overlord
juju deploy ch:hello-juju --revision=3
```

### CODE REVIEW

File: core/charm/repository/charmhub.go [hunk 1]
```diff
@@ -124,6 +124,14 @@ func (c *CharmHubRepository) ResolveWithPreferredChannel(charmURL *charm.URL, re
 		// specified channel for future refresh calls, is located in the origin.
 		if res.Entity.Type == transport.CharmType && requestedOrigin.Channel != nil {
 			effectiveChannel = requestedOrigin.Channel.String()
+			if effectiveChannel == "" {
+				// Tell the user they need to specify a channel
+				return nil, corecharm.Origin{}, nil, errors.Errorf(
```
Comment (tlm) [inline core/charm/repository/charmhub.go:129]: errors.New instead of errors.Errorf

**Comment by barrettj12** [comment]
@tlm pointed out that we should not be giving instructions about the CLI inside Juju Core.

Instead (if I understood correctly):
- raise a "channel empty" error inside Juju core
- send this error up the stack
- then convert(?) to an error code which can be sent to the CLI
- once received by the CLI, convert to human-readable message to print

I need to make these changes before this PR is merged.

**Comment by barrettj12** [comment]
The error was already being caught on the controller side, so I just added a check on the client side.

**Comment by tlm** [comment]
!!build!!

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -720,7 +720,16 @@ func (c *DeployCommand) Init(args []string) error {
 		// do a late validation at Run().
 		c.unknownModel = true
 	}
-	if c.channelStr != "" {
+	if c.channelStr == "" {
```
Comment (tlm) [inline cmd/juju/application/deploy.go:723]: Combine these 2 lines together

```go
if c.channelStr == "" && c.Revision != -1 {
```

File: cmd/juju/application/deploy.go [hunk 3]
```diff
@@ -720,7 +720,16 @@ func (c *DeployCommand) Init(args []string) error {
...
+		if c.Revision != -1 {
+			// Tell the user they need to specify a channel
+			return errors.BadRequestf(
```
Comment (tlm) [inline cmd/juju/application/deploy.go:726]: I would drop Bad Request here and just return a plane old Errors.New

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -720,7 +720,16 @@ func (c *DeployCommand) Init(args []string) error {
...
+				"If a revision is specified, a channel must also be specified for store\n" +
```
Comment (tlm) [inline cmd/juju/application/deploy.go:727]: Might read better to make this as a go literal string with ``

File: cmd/juju/application/deploy.go [hunk 5]
```diff
@@ -720,7 +720,16 @@ func (c *DeployCommand) Init(args []string) error {
...
+					"charms and bundles. Do this using the --channel option, e.g.\n" +
+					"    juju deploy <charm> --revision=5 --channel=stable",
+			)
+		}
+	} else {
```
Comment (tlm) [inline cmd/juju/application/deploy.go:732]: Break out as it's own if statement as the previous one has a return in it.

Comment (hpidcock) [reply cmd/juju/application/deploy.go:732]: Here is a few points/questions that can help make code cleaner:
- Exit early.
- Prefer long functions over deep functions.
- Would inverting the if condition make the function simpler (i.e. longer over deeper)?
- Can you make the function shallower (less nested scopes) by moving a block to another function?

**Review by tlm — APPROVED** [review]

**Comment by tlm** [comment]
$$merge$$

**Pull request merged**